### ISSUE TITLE
train_loss_epoch=nan.0

### ISSUE THREAD

**Issue by Escape142** [sb-ai-lab/RePlay#90]
## 🐛 Bug

`SasRec` training becomes numerically unstable with `bf16-mixed` precision and the training loss eventually turns into `NaN`. The same training setup works correctly with `fp32`.

This looks related to mixed-precision training in the sampled loss path. I see the issue in my current training pipeline when using `bf16-mixed`, while `32-true` runs normally.

### To Reproduce

1. Train `SasRec` with Lightning using:
   - `precision="bf16-mixed"`
   - sampled loss / negative sampling enabled
2. Start training on GPU
3. Observe that training initially progresses normally, but later `train_loss_step` becomes `nan`

Example setup:
- PyTorch Lightning trainer with `precision="bf16-mixed"`
- GPU training
- `SasRec` model
- CE loss

Example log output:

```text
Using bfloat16 Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [3]

  | Name  | Type   | Params | Mode
-----------------------------------------
0 | model | SasRec | 4.3 M  | train
-----------------------------------------
4.3 M     Trainable params
0         Non-trainable params
4.3 M     Total params
17.126    Total estimated model params size (MB)

Epoch 0:  71%|██████████████████████████████████████████████████████████████████████▏
| 5880/8295 [04:01<01:39, 24.37it/s, v_num=0, learning_rate_step=0.0005, train_loss_step=nan.0

**Comment by OnlyDeniko** [comment]
Hi @Escape142!  Thanks for the report and for sharing the environment details.

We were not able to reproduce this issue on our side when running the same setup on **NVIDIA A100**.

Could you please try the following checks:

1. Run the same training configuration with `precision="32-true"` (same data/config/seed if possible).
2. If training is stable in FP32, that strongly suggests the issue is related to mixed precision (`bf16-mixed` / AMP).
3. Try enabling gradient clipping in the Trainer, for example:
   - `gradient_clip_val=1.0`
   - `gradient_clip_algorithm="norm"`
4. If needed, try a smaller clip value (e.g. `0.5`) and/or reduce learning rate (e.g. 2x).

Note: gradient clipping helps when NaNs come from exploding gradients (after backward). If NaN appears already in forward/loss computation, clipping alone may not be sufficient.

If you can share results from the FP32 run and clipping test, we can narrow this down quickly.

**Comment by Escape142** [comment]
Updated PyTorch to 2.5.1 and the training runs without NaNs now, so this issue can be closed.
